### ISSUE TITLE
Nullable reference types: Internal assertion failures

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#9933]
I have been playing with nullable reference types, which is already working remarkably great! To try it out, I have [patched up](https://github.com/bkoelman/ResharperCodeContractNullability/tree/staticnull) my analyzer to convert Resharper's `(Item)NotNull/CanBeNullAttribute` annotations to nullable reference types. Then ran that on an [actual annotated codebase](https://github.com/bkoelman/DogAgilityCompetitionManagement). 

During that process I discovered some oddities, which I'll create issues for here.

**Version Used**: Built from source, branch NullableReferenceTypes (commit https://github.com/dotnet/roslyn/commit/8553cdac75c2b0901e8ae401aaa23022f0689935)

**Steps to Reproduce**:
When uncommenting the line in the next code snippet, Roslyn internal assertions are raised.

``` csharp
using System.Threading;
using System.Threading.Tasks;

namespace DogAgilityCompetition.Circe.Session
{
    public sealed class ActionQueue
    {
        private struct WorkItem
        {
            public TaskCompletionSource<object> TaskSource { get; }

            // Uncommenting the next line raises a sequence of Roslyn internal assertion failures
            //public CancellationToken CancelToken { get; }

            public WorkItem(int some)
                : this()
            {
            }
        }
    }
}
```

**Expected Behavior**: No assertions that fail.

**Actual Behavior**:
First failed assertion occurs at https://github.com/dotnet/roslyn/blob/features/NullableReferenceTypes/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs#L1393 (value of `targetSlot`: -1). At that moment, `targetType` parameter is `System.Threading.CancellationToken`. One frame up the call stack (method `InheritNullableStateOfFieldOrProperty`), parameter `fieldOrProperty` displays in debugger: `Field System.Threading.CancellationToken DogAgilityCompetition.Circe.Session.ActionQueue.WorkItem.<CancelToken>k__BackingField`.

Hope that helps!

@AlekseyTs
